// remotion/dist/cjs/resolve-video-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveVideoConfig = void 0;
const validate_default_codec_js_1 = require("./validation/validate-default-codec.js");
const validate_dimensions_js_1 = require("./validation/validate-dimensions.js");
const validate_duration_in_frames_js_1 = require("./validation/validate-duration-in-frames.js");
const validate_fps_js_1 = require("./validation/validate-fps.js");
const validateCalculated = ({ composition, calculated, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const calculateMetadataErrorLocation = `calculated by calculateMetadata() for the composition "${composition.id}"`;
    const defaultErrorLocation = `of the "<Composition />" component with the id "${composition.id}"`;
    const width = (_b = (_a = calculated === null || calculated === void 0 ? void 0 : calculated.width) !== null && _a !== void 0 ? _a : composition.width) !== null && _b !== void 0 ? _b : undefined;
    (0, validate_dimensions_js_1.validateDimension)(width, 'width', (calculated === null || calculated === void 0 ? void 0 : calculated.width) ? calculateMetadataErrorLocation : defaultErrorLocation);
    const height = (_d = (_c = calculated === null || calculated === void 0 ? void 0 : calculated.height) !== null && _c !== void 0 ? _c : composition.height) !== null && _d !== void 0 ? _d : undefined;
    (0, validate_dimensions_js_1.validateDimension)(height, 'height', (calculated === null || calculated === void 0 ? void 0 : calculated.height) ? calculateMetadataErrorLocation : defaultErrorLocation);
    const fps = (_f = (_e = calculated === null || calculated === void 0 ? void 0 : calculated.fps) !== null && _e !== void 0 ? _e : composition.fps) !== null && _f !== void 0 ? _f : null;
    (0, validate_fps_js_1.validateFps)(fps, (calculated === null || calculated === void 0 ? void 0 : calculated.fps) ? calculateMetadataErrorLocation : defaultErrorLocation, false);
    const durationInFrames = (_h = (_g = calculated === null || calculated === void 0 ? void 0 : calculated.durationInFrames) !== null && _g !== void 0 ? _g : composition.durationInFrames) !== null && _h !== void 0 ? _h : null;
    (0, validate_duration_in_frames_js_1.validateDurationInFrames)(durationInFrames, {
        allowFloats: false,
        component: `of the "<Composition />" component with the id "${composition.id}"`,
    });
    const defaultCodec = calculated === null || calculated === void 0 ? void 0 : calculated.defaultCodec;
    (0, validate_default_codec_js_1.validateDefaultCodec)(defaultCodec, calculateMetadataErrorLocation);
    return { width, height, fps, durationInFrames, defaultCodec };
};
const resolveVideoConfig = ({ composition, editorProps: editorPropsOrUndefined, signal, inputProps, }) => {
    var _a, _b, _c, _d, _e, _f;
    const originalProps = {
        ...((_a = composition.defaultProps) !== null && _a !== void 0 ? _a : {}),
        ...(editorPropsOrUndefined !== null && editorPropsOrUndefined !== void 0 ? editorPropsOrUndefined : {}),
        ...(inputProps !== null && inputProps !== void 0 ? inputProps : {}),
    };
    const calculatedProm = composition.calculateMetadata
        ? composition.calculateMetadata({
            defaultProps: (_b = composition.defaultProps) !== null && _b !== void 0 ? _b : {},
            props: originalProps,
            abortSignal: signal,
            compositionId: composition.id,
        })
        : null;
    if (calculatedProm !== null &&
        typeof calculatedProm === 'object' &&
        'then' in calculatedProm) {
        return calculatedProm.then((c) => {
            var _a, _b;
            const { height, width, durationInFrames, fps, defaultCodec } = validateCalculated({
                calculated: c,
                composition,
            });
            return {
                width,
                height,
                fps,
                durationInFrames,
                id: composition.id,
                defaultProps: (_a = composition.defaultProps) !== null && _a !== void 0 ? _a : {},
                props: (_b = c.props) !== null && _b !== void 0 ? _b : originalProps,
                defaultCodec: defaultCodec !== null && defaultCodec !== void 0 ? defaultCodec : null,
            };
        });
    }
    const data = validateCalculated({
        calculated: calculatedProm,
        composition,
    });
    if (calculatedProm === null) {
        return {
            ...data,
            id: composition.id,
            defaultProps: (_c = composition.defaultProps) !== null && _c !== void 0 ? _c : {},
            props: originalProps,
            defaultCodec: null,
        };
    }
    return {
        ...data,
        id: composition.id,
        defaultProps: (_d = composition.defaultProps) !== null && _d !== void 0 ? _d : {},
        props: (_e = calculatedProm.props) !== null && _e !== void 0 ? _e : originalProps,
        defaultCodec: (_f = calculatedProm.defaultCodec) !== null && _f !== void 0 ? _f : null,
    };
};
exports.resolveVideoConfig = resolveVideoConfig;
