// @remotion/renderer/dist/options/color-space.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateColorSpace = exports.colorSpaceOption = exports.validColorSpaces = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
exports.validColorSpaces = ['default', 'bt709', 'bt2020-ncl'];
let colorSpace = 'default';
const cliFlag = 'color-space';
exports.colorSpaceOption = {
    name: 'Color space',
    cliFlag: 'color-space',
    description: () => ((0, jsx_runtime_1.jsxs)(jsx_runtime_1.Fragment, { children: ["Color space to use for the video. Acceptable values:", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "default", '"'] }), ",", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "bt709", '"'] }), ' ', "(since v4.0.28),", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "bt2020-ncl", '"'] }), ' ', "(since v4.0.88),", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "bt2020-cl", '"'] }), ' ', "(since v4.0.88), .", (0, jsx_runtime_1.jsx)("br", {}), "If a non-default colorspace is used, it is recommended to also use", ' ', (0, jsx_runtime_1.jsxs)("code", { children: ['"', "png", '"'] }), ' ', "as the image format to have accurate color transformations throughout. Only since v4.0.83, colorspace conversion is actually performed, previously it would only tag the metadata of the video."] })),
    docLink: 'https://www.remotion.dev/docs/renderer/render-media#colorspace',
    ssrName: 'colorSpace',
    type: 'default',
    getValue: ({ commandLine }) => {
        if (commandLine[cliFlag] !== undefined) {
            return {
                source: 'cli',
                value: commandLine[cliFlag],
            };
        }
        if (colorSpace !== 'default') {
            return {
                source: 'config',
                value: colorSpace,
            };
        }
        return {
            source: 'default',
            value: 'default',
        };
    },
    setConfig: (value) => {
        colorSpace = value;
    },
};
const validateColorSpace = (option) => {
    if (exports.validColorSpaces.includes(option)) {
        return;
    }
    throw new Error(`Expected one of ${exports.validColorSpaces.map((c) => `"${c}"`).join(', ')} for ${exports.colorSpaceOption.ssrName} but got "${option}"`);
};
exports.validateColorSpace = validateColorSpace;
