// @remotion/studio/dist/components/ZoomPersistor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoomPersistor = exports.deriveCanvasContentFromUrl = exports.getZoomFromLocalStorage = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const url_state_1 = require("../helpers/url-state");
const timeline_zoom_1 = require("../state/timeline-zoom");
const makeKey = () => {
    return `remotion.zoom-map`;
};
const persistCurrentZoom = (zoom) => {
    localStorage.setItem(makeKey(), JSON.stringify(zoom));
};
const getZoomFromLocalStorage = () => {
    const zoom = localStorage.getItem(makeKey());
    return zoom ? JSON.parse(zoom) : {};
};
exports.getZoomFromLocalStorage = getZoomFromLocalStorage;
const deriveCanvasContentFromUrl = () => {
    const substrings = (0, url_state_1.getRoute)().split('/').filter(Boolean);
    const lastPart = substrings[substrings.length - 1];
    if (substrings[0] === 'assets') {
        return {
            type: 'asset',
            asset: decodeURIComponent((0, url_state_1.getRoute)().substring('/assets/'.length)),
        };
    }
    if (substrings[0] === 'outputs') {
        return {
            type: 'output',
            path: decodeURIComponent((0, url_state_1.getRoute)().substring('/outputs/'.length)),
        };
    }
    if (lastPart) {
        return {
            type: 'composition',
            compositionId: lastPart,
        };
    }
    return null;
};
exports.deriveCanvasContentFromUrl = deriveCanvasContentFromUrl;
const ZoomPersistor = () => {
    const [playing] = remotion_1.Internals.Timeline.usePlayingState();
    const { zoom } = (0, react_1.useContext)(timeline_zoom_1.TimelineZoomCtx);
    const { canvasContent } = (0, react_1.useContext)(remotion_1.Internals.CompositionManager);
    const urlState = (0, exports.deriveCanvasContentFromUrl)();
    const isActive = urlState &&
        urlState.type === 'composition' &&
        canvasContent &&
        canvasContent.type === 'composition' &&
        urlState.compositionId === canvasContent.compositionId;
    (0, react_1.useEffect)(() => {
        if (!isActive) {
            return;
        }
        persistCurrentZoom(zoom);
    }, [zoom, isActive, playing, urlState]);
    return null;
};
exports.ZoomPersistor = ZoomPersistor;
