// @remotion/cli/dist/parse-command-line.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.quietFlagProvided = exports.parseCommandLine = exports.parsedCli = exports.BooleanFlags = void 0;
const client_1 = require("@remotion/renderer/client");
const minimist_1 = __importDefault(require("minimist"));
const config_1 = require("./config");
const log_1 = require("./log");
const { beepOnFinishOption, colorSpaceOption, offthreadVideoCacheSizeInBytesOption, encodingBufferSizeOption, encodingMaxRateOption, deleteAfterOption, folderExpiryOption, enableMultiprocessOnLinuxOption, numberOfGifLoopsOption, x264Option, enforceAudioOption, jpegQualityOption, audioBitrateOption, videoBitrateOption, audioCodecOption, publicPathOption, } = client_1.BrowserSafeApis.options;
exports.BooleanFlags = [
    'overwrite',
    'force',
    'sequence',
    'help',
    'quiet',
    'q',
    'muted',
    enforceAudioOption.cliFlag,
    // Lambda flags
    'force',
    'disable-chunk-optimization',
    'save-browser-logs',
    'disable-cloudwatch',
    'enable-lambda-insights',
    'yes',
    'y',
    'disable-web-security',
    'ignore-certificate-errors',
    'disable-headless',
    'disable-keyboard-shortcuts',
    'default-only',
    'no-open',
    'ipv4',
    beepOnFinishOption.cliFlag,
    'repro',
];
exports.parsedCli = (0, minimist_1.default)(process.argv.slice(2), {
    boolean: exports.BooleanFlags,
    default: {
        overwrite: true,
    },
});
const parseCommandLine = () => {
    if (exports.parsedCli['pixel-format']) {
        config_1.Config.setPixelFormat(exports.parsedCli['pixel-format']);
    }
    if (exports.parsedCli['browser-executable']) {
        config_1.Config.setBrowserExecutable(exports.parsedCli['browser-executable']);
    }
    if (typeof exports.parsedCli['bundle-cache'] !== 'undefined') {
        config_1.Config.setCachingEnabled(exports.parsedCli['bundle-cache'] !== 'false');
    }
    if (exports.parsedCli['disable-web-security']) {
        config_1.Config.setChromiumDisableWebSecurity(true);
    }
    if (exports.parsedCli['ignore-certificate-errors']) {
        config_1.Config.setChromiumIgnoreCertificateErrors(true);
    }
    if (exports.parsedCli['user-agent']) {
        config_1.Config.setChromiumUserAgent(exports.parsedCli['user-agent']);
    }
    if (exports.parsedCli.concurrency) {
        config_1.Config.setConcurrency(exports.parsedCli.concurrency);
    }
    if (exports.parsedCli.height) {
        config_1.Config.overrideHeight(exports.parsedCli.height);
    }
    if (exports.parsedCli.width) {
        config_1.Config.overrideWidth(exports.parsedCli.width);
    }
    if (exports.parsedCli.frames) {
        config_1.ConfigInternals.setFrameRangeFromCli(exports.parsedCli.frames);
    }
    if (exports.parsedCli.frame) {
        config_1.ConfigInternals.setStillFrame(Number(exports.parsedCli.frame));
    }
    if (exports.parsedCli.png) {
        throw new Error('The --png flag has been removed. Use --sequence --image-format=png from now on.');
    }
    if (exports.parsedCli.sequence) {
        config_1.Config.setImageSequence(true);
    }
    if (exports.parsedCli['every-nth-frame']) {
        config_1.Config.setEveryNthFrame(exports.parsedCli['every-nth-frame']);
    }
    if (exports.parsedCli['prores-profile']) {
        config_1.Config.setProResProfile(String(exports.parsedCli['prores-profile']));
    }
    if (typeof exports.parsedCli.quality !== 'undefined') {
        log_1.Log.warn({ indent: false, logLevel: 'info' }, 'The --quality flag has been renamed to --jpeg-quality instead.');
        config_1.Config.setJpegQuality(exports.parsedCli.quality);
    }
    if (typeof exports.parsedCli.scale !== 'undefined') {
        config_1.Config.setScale(exports.parsedCli.scale);
    }
    if (typeof exports.parsedCli['disable-keyboard-shortcuts'] !== 'undefined') {
        config_1.Config.setKeyboardShortcutsEnabled(!exports.parsedCli['disable-keyboard-shortcuts']);
    }
    if (typeof exports.parsedCli['webpack-poll'] !== 'undefined') {
        config_1.Config.setWebpackPollingInMilliseconds(exports.parsedCli['webpack-poll']);
    }
};
exports.parseCommandLine = parseCommandLine;
const quietFlagProvided = () => exports.parsedCli.quiet || exports.parsedCli.q;
exports.quietFlagProvided = quietFlagProvided;
